AT: update eligibility screen design

Copons opened the pull request:
Fixes https://github.com/Automattic/wp-calypso/issues/11394, https://github.com/Automattic/wp-calypso/issues/11486, and https://github.com/Automattic/wp-calypso/issues/11514

This PR aims at updating the eligibility screen design - for both plugins and themes flows - to be more clear and less scary.

The current logic for showing the upgrade banners follows this cascading checks:

- Has no Business Plan? Show an "Upgrade to Business Plan" banner.
- Has Business Plan but no custom domain? Show an "Add a free custom domain" banner.

Each banner is expected to come in two copy-flavours, for plugins and themes.
E.g. "Upgrade to Business Plan to install this plugin", or "Upgrade to Business Plan to upload this theme".

Of all the holds currently covered (see: https://github.com/Automattic/wp-calypso/blob/master/client/blocks/eligibility-warnings/index.jsx#L23-L71), only `NO_BUSINESS_PLAN` and `NOT_USING_CUSTOM_DOMAIN` are replaced with a banner.

## Future updates

Not currently included in this PR (possibly in a future one):

- Decide if more holds need their own banner.
- Update copy for banners, holds and warnings.
- Update holds' `supportUrl` (possibly using the same logic used for warnings?).

## Screenshots

### Before
<img width="811" alt="screen shot 2017-02-14 at 12 45 17 pm" src="https://cloud.githubusercontent.com/assets/1464705/22949959/d14a042e-f2b8-11e6-8cae-a6602012b2a0.png">

### After

**Non-Business, private**
<img width="686" alt="non-biz-private" src="https://cloud.githubusercontent.com/assets/2070010/23412524/b4747f9e-fdcd-11e6-98e2-513a4d4077a5.png">

**Non-Business, no warnings**
<img width="685" alt="non-biz-no-warnings" src="https://cloud.githubusercontent.com/assets/2070010/23412525/b475450a-fdcd-11e6-9ff9-71c600b567eb.png">

**Business, no domain**
<img width="685" alt="biz-no-domain" src="https://cloud.githubusercontent.com/assets/2070010/23412523/b4745046-fdcd-11e6-943f-91c266e5e7ef.png">

**No conflicts**
<img width="650" alt="screen shot 2017-03-06 at 12 17 05" src="https://cloud.githubusercontent.com/assets/2070010/23609752/768268f6-0267-11e7-82f4-4f22683eab80.png">
*(please don't mind the "upload plugins" typo!)*

folletto commented:
This would fix also the logic in #11514 right? Or it's a separate PR after this?

Copons commented:
@folletto Simply moving the banner would be ok for this PR, but eventually the whole logic for _which_ banner to show will radically change.

Does it make more sense to move it here (an almost design-only PR) or in a full-fledged banner-only PR?

folletto commented:
I feel that if this PR lands the feature for Plugins, it makes sense to land it on Themes too as well... they should be the same even if themes has an extra check to decide IF that has to be shown.

The only reason to not do that is if the size of the PR grows too much, but otherwise I'd try. :)

Copons commented:
@folletto I tried it just for fun and it was a fairly small change so I just pushed it.
Care to give it a try if I missed anything since I don't know exactly all that goes on in the theme uploads page?

a8ci18n commented on client/blocks/eligibility-warnings/index.jsx:
```diff
@@ -2,139 +2,71 @@
-			) }
+			{ 'plugins' === context && ! hasBusinessPlan && ! isJetpack &&
+				<Banner
+					description={ translate( 'Please upgrade to install this plugin.' ) }
```
Hi! I've found a possible matching string that has already been translated **17** times:
`translate( 'Upgrade to the Business plan to install plugins.' )` *ES Score: 10*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

folletto commented:
It's working well... except on an active Jetpack site where the form turns up disabled. I'm not sure why tho, as the form should be independent from the eligibility check, but maybe something cascaded down?

To replicate just open any Jetpack site and see the upload page:

https://calypso.live/design/upload/$jetpacksite?branch=update/plugins/at-eligibility-design

Copons commented:
@folletto Ah yup. I removed too many checks from theme upload, without noticing they were used somewhere else too.
Should be ok now.

folletto commented:
Checked:

* Jetpack → Themes → Upload — ✅ 
* .com without Business plan → Themes → Upload — ✅ 
* .com with Business plan → Themes → Upload — ✅ 

All's good there. :)

---

One thing: I'm wondering if the icons (i.e. red ❌  ) should be aligned to the top line of the block, not vertically in the block. That's the only design question I have, I defer to @drw158 to double check the change tho. :) 

Also: the text need a review, but let's push this in first with all the logic and the design, and let's open another one with the edited text (let's not close #11394 until the copy is updated too).

a8ci18n commented on client/blocks/eligibility-warnings/index.jsx:
```diff
@@ -2,139 +2,85 @@
+					description={ translate( 'Add a free custom domain to install this plugin.' ) }
+					href={ `/domains/add/${ siteSlug }` }
+					icon="domains"
+					title={ translate( 'Custom domain required' ) }
```
Hi! I've found a possible matching string that has already been translated **27** times:
`translate( 'Custom Domain' )` *ES Score: 7*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/plugins/at-eligibility-design)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

Copons commented:
@folletto, cc @drw158 

> One thing: I'm wondering if the icons (i.e. red ❌ ) should be aligned to the top line of the block, not vertically in the block. That's the only design question I have, I defer to @drw158 to double check the change tho. :)

I'm not even sure that's a `cross` or `small-cross` to be honest.
Anyway, icon alignment is a matter of a CSS rule, so whatever looks better to you designers. :)

> Also: the text need a review, but let's push this in first with all the logic and the design, and let's open another one with the edited text (let's not close #11394 until the copy is updated too).

I was waiting to have a more complete code to ask for a copy review.
I guess now, after a PR description update, is a good time though.

Copons commented:
@folletto I've changed `cross` with `cross-small` and aligned everything to the top.
It looks better to me, but that's really up to you.

folletto commented:
> I've changed cross with cross-small and aligned everything to the top.
It looks better to me, but that's really up to you.

I prefer that approach too :) 

Design wise it's good. I know copy could use a review, but I'd suggest to ship this first as it's already a solid improvement, and review the text separately if it's ok for you :)

roundhill commented:
Could you also see if we can fix #11687 via this PR, if it doesn't already?

roundhill commented on client/blocks/eligibility-warnings/index.jsx:
```diff
@@ -2,139 +2,85 @@
+				<Banner
+					className="eligibility-warnings__banner"
+					description={ translate( 'Add a free custom domain to install this plugin.' ) }
+					href={ `/domains/add/${ siteSlug }` }
```
I think this link is incorrect, since most sites should already have a domain with the business plan.

`/domains/manage/${ siteSlug }`

drw158 replied:
Just a string change then? 

Claim your domain credit

Copons replied:
@roundhill If you don't have a custom domain, `/domains/manage` brings you to a page with a big blue notice saying "Free domain available - Claim Free Domain", and clicking on it - which would be the action required to solve the AT hold - opens `/domains/add`, where you can pick your custom domain of choice for free.

So I'd say we should keep `/domains/add` here, in order to save the user one click and a bit of confusion.

roundhill commented on client/blocks/eligibility-warnings/warning-list.jsx:
```diff
@@ -0,0 +1,54 @@
+	return (
+		<div>
+			<SectionHeader label={ translate(
+				"By proceeding you'll lose %d feature:",
```
Do we explain somewhere else _why_ they will lose these features? If not, maybe something like:

`In order for your site to support this plugin you'll lose %d feature:`

Copons commented:
@roundhill Since it's likely we'll have to think about and change a lot of copy and links (see: the holds links are still empty), I believe it'd better to do it all in a follow up PR.

folletto commented:
I agree. This is mostly about the layout, let's push this in, then review copy. It's better to break down these kinds of PRs that are improvements by themselves and chain more PRs as follow ups (referencing all of them to the same original issue also helps with discovery).

roundhill commented:
> I believe it'd better to do it all in a follow up PR.

Sounds good 👍

a8ci18n commented on client/blocks/eligibility-warnings/index.jsx:
```diff
@@ -2,144 +2,99 @@
+
+			{ 'plugins' === context && ! hasBusinessPlan && ! isJetpack &&
+				<Banner
+					description={ translate( 'Please upgrade to install this plugin.' ) }
```
Hi! I've found a possible matching string that has already been translated **18** times:
`translate( 'Upgrade to the Business plan to install plugins.' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

Copons commented:
@roundhill: regarding #11687, I've added a `Notice` saying everything's good, and added more checks to make sure no warning messages are displayed anywhere when there are no conflicts.

![https://cloud.githubusercontent.com/assets/2070010/23557466/13a36a50-0028-11e7-8423-d18673871472.png](https://cloud.githubusercontent.com/assets/2070010/23557466/13a36a50-0028-11e7-8423-d18673871472.png)

As usual: copy is a placeholder waiting for the follow up PR tackling it.

cc: @folletto @drw158 for a design opinion

folletto commented:
If there are no conflict the UI should just not show up and proceed with whichever page should be showing up.

I've read #11687 but it's not clear to me in which case that screen would be shown for plugins... I mean I get accessing a specific page for testing, but the user shouldn't see that for plugins (as the install should show the message first) and neither for themes (which would simply show the upload form). What's the flow that gets me there?

Copons commented:
@folletto I'm assuming it can happen if you get to the eligibility warnings screen, then go fix the conflicts in other tabs. Then you go back to the eligibility page and hit refresh.

At that point, ideally, you would simply trigger the AT without showing anything.
But there's the possibility that you are redirected to that page (say, via a "malicious" shortened url, or simply by accident) without the intention of actually AT-ing.
If that happens, and you have no conflicts, we shouldn't just transfer your site without notice.

I mean, it is somewhat of an edge case but can definitely happen, and I believe it's better to err on the side of caution.

folletto commented:
> I'm assuming it can happen if you get to the eligibility warnings screen, then go fix the conflicts in other tabs. Then you go back to the eligibility page and hit refresh.

Ok clear. In theory "going back to the eligibility page" should go through either the Plugin page or the Theme page too, so this should end up a fairly edge case. 

Which means the flows are:

1. Themes → Upload (show eligibility) → fix + fix + fix → Themes → Upload (show upload)
2. Plugins (Add) → Install button (if not "Upgrade to Business" Banner) → Notice → Eligibility screen → fix + fix + fix → Plugins (Add) → Install

In practice only Plugins should thus show the eligibility screen on reload as for Themes it's the same page as Upload. 

Which means that there's no need to automate the installation, as if the user refreshes that eligibility page it's a manual choice they did instead of trying again to install the plugin.

This leads to two things:

1. The message shouldn't be a `Notice` but a normal card with text inside. It's not a temporary notification, it's a message. 
2. The "Resolve" buttons on the eligibility page shouldn't open on a new tab (but they should be openable on a new tab – i.e. with `href` → right click should work).

Here's a mock:

![screen shot 2017-03-06 at 10 51 09](https://cloud.githubusercontent.com/assets/4389/23606932/d9b35bb8-025a-11e7-9d8f-401c74d7b6ae.png)

I think the two fixes above should solve the issue. Let's merge this PR as soon as we can. :)

a8ci18n commented on client/blocks/eligibility-warnings/index.jsx:
```diff
@@ -2,144 +2,100 @@
+
+			{ 'plugins' === context && ! hasBusinessPlan && ! isJetpack &&
+				<Banner
+					description={ translate( 'Please upgrade to install this plugin.' ) }
```
Hi! I've found a possible matching string that has already been translated **18** times:
`translate( 'Upgrade to the Business plan to install plugins.' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

Copons commented:
@folletto Pushed both fixes.

This is the updated no-conflicts screen, which, well, looks pretty much like yours :)
<img width="650" alt="screen shot 2017-03-06 at 12 17 05" src="https://cloud.githubusercontent.com/assets/2070010/23609790/ad0a9722-0267-11e7-8ae4-c8e078fbcb69.png">

I've removed the Resolve buttons' `target="_blank"`, but I've kept it for the features warnings `❓`, since they open support pages (sort of external) as a reference/explanation for the features the user's gonna lose.

folletto commented:
Ok, sounds good! 👍

dmsnell commented on client/blocks/eligibility-warnings/hold-list.jsx:
```diff
@@ -0,0 +1,99 @@
+			) } />
+			<Card className="eligibility-warnings__hold-list">
+				{ map( holds, ( hold, index ) =>
+					<div className="eligibility-warnings__hold" key={ index }>
```
danger! list-index as React key - can we use a property which identifies the hold here? we probably have no reasonable guarantee that this list won't reorder

dmsnell commented on client/blocks/eligibility-warnings/index.jsx:
```diff
@@ -2,144 +2,100 @@
+	siteSlug,
+	translate,
+} ) => {
+	const context = -1 !== backUrl.indexOf( 'plugins' ) ? 'plugins' : 'themes';
```
please feel free to use `lodash.includes()` here to better indicate the goal: "does `backUrl` contain the string 'plugins'

dmsnell commented on client/blocks/eligibility-warnings/index.jsx:
```diff
@@ -2,144 +2,100 @@
-	} );
+	let holds = get( eligibilityData, 'eligibilityHolds', [ 'PLACEHOLDER', 'PLACEHOLDER' ] );
+	const bannerHolds = filter( holds, hold => -1 !== [ 'NO_BUSINESS_PLAN', 'NOT_USING_CUSTOM_DOMAIN' ].indexOf( hold ) );
+	holds = difference( holds, bannerHolds );
```
do we have a reason to mutate this so quickly instead of just declaring it? it's an odd use of `filter` and `difference` - did we mean to use a different set function? what's the goal here? to remove the banner holds from the list of holds?

Copons replied:
Yup, that's pretty much it.
Some holds are supposed to show as banners, some as list items.
I suppose I could rewrite it as
```js
const holds = get( ... );
const bannerHolds = filter( holds, ... );
const listHolds = difference( holds, bannerHolds );
```

dmsnell replied:
so if our goal is to split the list into two lists based on some predicate we have a handy method!

```js
const [ listHolds, bannerHolds ] = partition(
	get( eligibilityData, 'eligibilityHolds', [ 'PLACEHOLDER', 'PLACEHOLDER' ] ), // list of possible holds
	hold => includes( [ 'NO_BUSINESS_PLAN', 'NOT_USING_CUSTOM_DOMAIN' ], hold ), // these two are banner-only holds
);
```

Copons replied:
This looks swell!

a8ci18n commented on client/blocks/eligibility-warnings/hold-list.jsx:
```diff
@@ -0,0 +1,129 @@
+		},
+		NOT_RESOLVING_TO_WPCOM: {
+			title: translate( 'Primary domain not pointing to WordPress.com servers' ),
+			description: translate( 'Your primary domain must point to WordPress.com servers.' ),
```
:ok: This change is safe, we'll keep the existing 17 translations.

a8ci18n commented on client/blocks/eligibility-warnings/hold-list.jsx:
```diff
@@ -0,0 +1,129 @@
+		},
+		NO_JETPACK_SITES: {
+			title: translate( 'Jetpack site' ),
+			description: translate( 'This feature is not supported on Jetpack sites.' ),
```
Hi! I've found a possible matching string that has already been translated **2** times:
`translate( 'This feature is not supported on VIP sites.' )` *ES Score: 9*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/plugins/at-eligibility-design)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/blocks/eligibility-warnings/hold-list.jsx:
```diff
@@ -0,0 +1,129 @@
+			supportUrl: 'https://support.wordpress.com/'
+		},
+		NO_JETPACK_SITES: {
+			title: translate( 'Jetpack site' ),
```
Hi! I've found a possible matching string that has already been translated **16** times:
`translate( 'Manage Jetpack Sites' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/blocks/eligibility-warnings/hold-list.jsx:
```diff
@@ -0,0 +1,129 @@
+			supportUrl: '',
+		},
+		TRANSFER_ALREADY_EXISTS: {
+			title: translate( 'Installation in progress' ),
```
Hi! I've found a possible matching string that has already been translated **31** times:
`translate( 'Installation', { context: 'Navigation item'} )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

roundhill commented:
This is looking good to me. 👍
